bagua-core-internal/src/comm_ops/decentralized_full_precision_asynchronous.rs

```diff
@@ -0,0 +1,133 @@
+use crate::comm_ops::decentralized_full_precision_synchronous::PeerSelectionMode;
+use crate::comm_ops::CommOpTrait;
+use crate::communicators::BaguaCommunicator;
+use crate::datatypes::{BaguaBucket, BaguaReductionOp, BaguaTensorRaw, RawBaguaTensor};
+use crate::resource_pool::CUDA_DEVICE_MEMORY_POOL;
+use crate::{BaguaCommOpChannels, BaguaCoreError};
+use crate::events::BaguaEventChannel;
+use std::sync::Arc;
+use std::time::Duration;
+use cuda_runtime_sys::{cudaStream_t, cudaEvent_t};
+
+
+#[derive(Debug)]
+pub struct DecentralizedFullPrecisionAsynchronous {
+    pub communicator: BaguaCommunicator,
+    pub peer_selection_mode: PeerSelectionMode,
+    pub torch_stream: u64,
+    pub src_ready_event: u64,
+    pub dst_ready_event: u64,
+}
+
+impl CommOpTrait for DecentralizedFullPrecisionAsynchronous {
+
+    fn execute_background_communication(
+            &self,
+            bucket: Arc<BaguaBucket>,
+            comm_op_channels: &BaguaCommOpChannels,
+    ) {
+        let bucket_guard = bucket.inner.lock();
+        let comm_stream = self.communicator.stream_ptr();
+
+        let mut communication_tensor = match &self.communicator {
+            BaguaCommunicator::SingleCommunicator(_) => {
+                bucket_guard.get_communication_tensor(comm_stream, false, false)
+            }
+            BaguaCommunicator::HierarchicalCommunicator(x) => {
+                panic!("asynchronous op only accepts non-hierarchical communicator");
+            },
+        };
+
+        let peer_mode = &self.peer_selection_mode;
```

Comment:
defined but never used.

better to assert the value (for example, currently the algorithm only supports "all"), and say that in the doc

Reply:
removed